src/lib.rs

```diff
@@ -369,10 +422,82 @@ pub unsafe extern "C" fn path_readlink(
     path_ptr: *const u8,
     path_len: usize,
     buf: *mut u8,
-    _buf_len: Size,
+    buf_len: Size,
     bufused: *mut Size,
 ) -> Errno {
-    unreachable()
+    let fd = wasi_filesystem::Descriptor::from_raw(fd as _);
+
+    let path = slice::from_raw_parts(path_ptr, path_len);
+
+    // If the user gave us a buffer shorter than `PATH_MAX`, it may not be
+    // long enough to accept the actual path. `cabi_realloc` can't fail,
+    // so instead we handle this case specially.
+    if buf_len < PATH_MAX {
+        return path_readlink_slow(fd, path, buf, buf_len, bufused);
+    }
+
+    register_buffer(buf, buf_len);
+
+    let result = fd.readlink_at(path);
+
+    unregister_buffer(buf, buf_len);
```

Comment:
Due to the `replace_*` calls in `cabi_realloc` I think this might always trap?

Reply:
Oh, good point. We can just remove that then.